File: python_backend/app/services/landmark_routing_policy.py
"""Landmark MODULE routing tiers — mirrors Java LandmarkRoutingPolicy."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import List, Optional


class RoutingTier(str, Enum):
    NONE = "none"
    SUGGESTED = "suggested"
    MANDATORY = "mandatory"


@dataclass(frozen=True)
class RoutingDecision:
    module_id: str
    tier: RoutingTier
    reason: str

    @property
    def applies(self) -> bool:
        return bool(self.module_id) and self.tier != RoutingTier.NONE


# 用户明确拒绝地标 MODULE（优先于别名子串匹配，避免「不要鸟巢」误命中「鸟巢」）
LANDMARK_REJECTION_MARKERS: List[str] = [
    "不要地标", "不要鸟巢", "不要仿", "不要复制", "别照搬", "非地标",
    "not landmark", "not a landmark", "no landmark",
    "don't copy", "do not copy",
]

# 多样化/原创措辞——仅在没有指名地标时，阻止 typology SUGGESTED 联想
VARIATION_INTENT_MARKERS: List[str] = [
    "原创", "独创", "独特", "不一样", "不要一样", "每次不同", "创新", "想象", "自由发挥",
    "自行设计", "自己设计",
    "generic", "original", "unique", "creative", "imaginative", "custom design",
    "varied", "different each time",
]

# 兼容旧 API：拒绝 + 多样化均视为「非强制路由」信号
CREATIVE_INTENT_MARKERS: List[str] = LANDMARK_REJECTION_MARKERS + VARIATION_INTENT_MARKERS

BIRDS_NEST_EXPLICIT: List[str] = [
    "鸟巢", "鸟巢体育馆", "国家体育场", "北京鸟巢",
    "bird's nest", "birds nest", "birds' nest", "beijing national stadium",
]

GOTHIC_CATHEDRAL_EXPLICIT: List[str] = [
    "哥特大教堂", "哥特式大教堂", "gothic cathedral",
]

MINGQING_COURTYARD_EXPLICIT: List[str] = [
    "明清官式院落", "明清官式建筑群落", "明清四合院",
    "ming qing courtyard", "mingqing courtyard",
]

CASTLE_COMPOUND_EXPLICIT: List[str] = [
    "中世纪城堡", "城堡复合体",
    "medieval castle compound", "medieval castle", "castle compound",
]

STADIUM_TYPOLOGY: List[str] = ["体育场", "体育馆", "stadium", "arena", "球场"]
ELLIPSE_TYPOLOGY: List[str] = ["椭圆", "椭圆形", "elliptical", "oval", "碗状", "看台"]

NAMED_ARCHITECT_MARKERS: List[str] = [
    "扎哈", "zaha", "哈迪德", "hadid",
    "贝聿铭", "i.m. pei", "pei",
    "安藤忠雄", "tadao ando",
    "诺曼·福斯特", "norman foster",
    "伦佐·皮亚诺", "renzo piano",
    "让·努维尔", "jean nouvel",
]

VARIATION_HINTS: List[str] = [
    "Vary facade_profile, entrance count, roof_type, masses[] offsets each generation.",
    "For MODULE: vary designSeed (1-9999), facing, bowlSteepness (0.25-0.45), dimensions hints.",
    "For 原创/独特 intent: use compositional MASS + PAVING + ROOF, not identical MODULE clones.",
]


def _contains_any(text: str, markers: List[str]) -> bool:
    lower = text.lower()
    return any(m.lower() in lower for m in markers)


def rejects_landmark_module(text: str) -> bool:
    if not (text or "").strip():
        return False
    return _contains_any(text, LANDMARK_REJECTION_MARKERS)


def is_variation_intent(text: str) -> bool:
    if not (text or "").strip():
        return False
    return _contains_any(text, VARIATION_INTENT_MARKERS)


def is_creative_or_original_intent(text: str) -> bool:
    """拒绝地标 或 仅要原创/多样化（无指名地标时阻止 SUGGESTED typology）。"""
    return rejects_landmark_module(text) or is_variation_intent(text)


def resolve_for_user_intent(text: str) -> Optional[RoutingDecision]:
    """
    指名地标（MANDATORY）优先于多样化措辞；
    多样化/拒绝仅在没有指名地标时阻止 typology SUGGESTED。
    """
    q = (text or "").strip()
    if not q:
        return None

    if rejects_landmark_module(q):
        return None

    lower = q.lower()

    # 指名鸟巢 — 即使同时含「不一样」等多样化词，仍强制 MODULE
    if _contains_any(lower, BIRDS_NEST_EXPLICIT):
        return RoutingDecision("birds_nest_stadium", RoutingTier.MANDATORY, "explicit_birds_nest")

    if _contains_any(lower, GOTHIC_CATHEDRAL_EXPLICIT) and not any(
        x in lower for x in ("notre dame", "巴黎圣母院", "cologne", "科隆", "chartres", "sagrada", "圣家族")
    ):
        return RoutingDecision("gothic_cathedral", RoutingTier.MANDATORY, "explicit_gothic_cathedral")

    if _contains_any(lower, MINGQING_COURTYARD_EXPLICIT):
        return RoutingDecision("mingqing_courtyard", RoutingTier.MANDATORY, "explicit_mingqing_courtyard")

    if _contains_any(lower, CASTLE_COMPOUND_EXPLICIT) and not any(
        x in lower for x in ("disney", "迪士尼", "himeji", "姬路", "japanese castle", "日式")
    ):
        return RoutingDecision("castle_compound", RoutingTier.MANDATORY, "explicit_castle_compound")

    # 点名建筑师 + 体育场 → 参数化形体，禁止默认 birds_nest typology
    if _contains_any(lower, NAMED_ARCHITECT_MARKERS) and _contains_any(lower, STADIUM_TYPOLOGY):
        return None

    # 无指名地标时的原创/多样化 → 不做 typology 联想
    if is_variation_intent(q):
        return None

    stadium = _contains_any(lower, STADIUM_TYPOLOGY)
    elliptical = _contains_any(lower, ELLIPSE_TYPOLOGY)
    if stadium and elliptical:
        return RoutingDecision(
            "birds_nest_stadium", RoutingTier.SUGGESTED, "typological_elliptical_stadium"
        )
    if stadium and ("现代" in lower or "modern" in lower):
        return RoutingDecision(
            "birds_nest_stadium", RoutingTier.SUGGESTED, "typological_modern_stadium"
        )

    return None


def variation_context_block() -> str:
    lines = ["[BUILDING VARIATION] (apply to every plan — users expect distinct results)"]
    lines.extend(f"- {h}" for h in VARIATION_HINTS)
    return "\n".join(lines)

File: src/main/java/com/formacraft/common/archetype/LandmarkRoutingPolicy.java
package com.formacraft.common.archetype;

import java.util.List;
import java.util.Locale;

/**
 * 地标 MODULE 路由分级：区分「指名地标」「类型相似」「原创/多样化」用户意图。
 * <p>
 * 同一风格与尺寸也应允许不同作品——只有用户明确点名地标时才强制 MODULE；
 * 类型匹配（如椭圆体育场）仅作推荐；原创/不要地标类措辞则跳过强制路由。
 */
public final class LandmarkRoutingPolicy {

    public enum RoutingTier {
        /** 不注入地标路由 */
        NONE,
        /** 推荐 MODULE，但允许 MASS/plan_program 等原创组合 */
        SUGGESTED,
        /** 用户指名地标，必须 MODULE */
        MANDATORY
    }

    public record RoutingDecision(String moduleId, RoutingTier tier, String reason) {
        public boolean applies() {
            return moduleId != null && !moduleId.isBlank() && tier != RoutingTier.NONE;
        }
    }

    private static final List<String> LANDMARK_REJECTION_MARKERS = List.of(
            "不要地标", "不要鸟巢", "不要仿", "不要复制", "别照搬", "非地标",
            "not landmark", "not a landmark", "no landmark",
            "don't copy", "do not copy"
    );

    private static final List<String> VARIATION_INTENT_MARKERS = List.of(
            "原创", "独创", "独特", "不一样", "不要一样", "每次不同", "创新", "想象", "自由发挥",
            "自行设计", "自己设计",
            "generic", "original", "unique", "creative", "imaginative", "custom design",
            "varied", "different each time"
    );

    /** @deprecated use {@link #rejectsLandmarkModule} / {@link #isVariationIntent} */
    @SuppressWarnings("unused")
    private static final List<String> CREATIVE_INTENT_MARKERS = List.of(
            "原创", "独创", "独特", "不一样", "不要一样", "每次不同", "创新", "想象", "自由发挥",
            "不要地标", "不要鸟巢", "不要仿", "不要复制", "别照搬", "非地标", "自行设计", "自己设计",
            "generic", "original", "unique", "creative", "imaginative", "custom design",
            "don't copy", "do not copy", "not landmark", "not a landmark", "one of a kind",
            "varied", "different each time", "no landmark"
    );

    /** 指名鸟巢/国家体育场——强制 MODULE */
    private static final List<String> BIRDS_NEST_EXPLICIT = List.of(
            "鸟巢", "鸟巢体育馆", "国家体育场", "北京鸟巢",
            "bird's nest", "birds nest", "birds' nest", "beijing national stadium"
    );

    /** 类型相似（椭圆/碗状体育场）——推荐 MODULE，允许原创路径 */
    private static final List<String> STADIUM_TYPOLOGY = List.of(
            "体育场", "体育馆", "stadium", "arena", "球场"
    );
    private static final List<String> ELLIPSE_TYPOLOGY = List.of(
            "椭圆", "椭圆形", "elliptical", "oval", "碗状", "看台"
    );

    private static final List<String> NAMED_ARCHITECT_MARKERS = List.of(
            "扎哈", "zaha", "哈迪德", "hadid",
            "贝聿铭", "i.m. pei", "pei",
            "安藤忠雄", "tadao ando",
            "诺曼·福斯特", "norman foster",
            "伦佐·皮亚诺", "renzo piano",
            "让·努维尔", "jean nouvel"
    );

    private LandmarkRoutingPolicy() {}

    public static boolean rejectsLandmarkModule(String text) {
        if (text == null || text.isBlank()) return false;
        String lower = text.toLowerCase(Locale.ROOT);
        for (String marker : LANDMARK_REJECTION_MARKERS) {
            if (lower.contains(marker.toLowerCase(Locale.ROOT))) {
                return true;
            }
        }
        return false;
    }

    public static boolean isVariationIntent(String text) {
        if (text == null || text.isBlank()) return false;
        String lower = text.toLowerCase(Locale.ROOT);
        for (String marker : VARIATION_INTENT_MARKERS) {
            if (lower.contains(marker.toLowerCase(Locale.ROOT))) {
                return true;
            }
        }
        return false;
    }

    public static boolean isCreativeOrOriginalIntent(String text) {
        return rejectsLandmarkModule(text) || isVariationIntent(text);
    }

    /**
     * 解析用户意图对应的地标路由决策；无匹配返回 {@code null}。
     * <p>指名地标（MANDATORY）优先于「不一样」等多样化措辞。</p>
     */
    public static RoutingDecision resolveForUserIntent(String text) {
        if (text == null || text.isBlank()) {
            return null;
        }
        if (rejectsLandmarkModule(text)) {
            return null;
        }

        String lower = text.toLowerCase(Locale.ROOT);

        if (containsAny(lower, BIRDS_NEST_EXPLICIT)) {
            return new RoutingDecision("birds_nest_stadium", RoutingTier.MANDATORY, "explicit_birds_nest");
        }

        // 其它地标：用户指名别名时强制 MODULE（优先于多样化措辞）
        for (LandmarkModuleRegistry.LandmarkModule module : LandmarkModuleRegistry.listModules()) {
            if ("birds_nest_stadium".equals(module.moduleId())) {
                continue;
            }
            if (mentionsExplicitLandmark(lower, module)) {
                return new RoutingDecision(module.moduleId(), RoutingTier.MANDATORY, "explicit_landmark");
            }
        }

        if (containsAny(lower, NAMED_ARCHITECT_MARKERS) && containsAny(lower, STADIUM_TYPOLOGY)) {
            return null;
        }

        if (isVariationIntent(text)) {
            return null;
        }

        boolean stadium = containsAny(lower, STADIUM_TYPOLOGY);
        boolean elliptical = containsAny(lower, ELLIPSE_TYPOLOGY);
        if (stadium && elliptical) {
            return new RoutingDecision("birds_nest_stadium", RoutingTier.SUGGESTED, "typological_elliptical_stadium");
        }
        if (stadium && (lower.contains("现代") || lower.contains("modern"))) {
            return new RoutingDecision("birds_nest_stadium", RoutingTier.SUGGESTED, "typological_modern_stadium");
        }

        return null;
    }

    /**
     * 注入 prompt 的「建筑多样化」原则——对所有 build 请求生效。
     */
    public static String promptVariationPrinciples() {
        return """

            ========================================
            BUILDING VARIATION (every user deserves a distinct result)
            ========================================
            Even with the same style_profile and similar dimensions, DO NOT output identical plans.
            - Vary facade_profile, entrance placement/count, roof_type, decorative_elements, masses[] offsets.
            - Vary style_attributes accents (materials, colors) within the requested style family.
            - For MODULE landmarks: still vary dimensions (within limits), params.facing, params.meshStructure,
              params.designSeed (integer), params.bowlSteepness (0.2–0.5), style_attributes materials.
            - Prefer compositional MASS + ROOF + PAVING + ENTRANCE when user wants 原创/独特/不要地标.
            - Only reproduce a canonical landmark form when the user explicitly names that landmark.

            """;
    }

    static String promptBlockForDecision(RoutingDecision decision) {
        if (decision == null || !decision.applies()) {
            return "";
        }
        String moduleId = decision.moduleId();
        if (decision.tier() == RoutingTier.MANDATORY) {
            return """

                ========================================
                LANDMARK MODULE ROUTING (MANDATORY FOR THIS REQUEST)
                ========================================
                User explicitly named landmark module: %s
                You MUST output exactly ONE component:
                  { "component_type": "MODULE",
                    "relative_position": {"x":0,"y":0,"z":0},
                    "dimensions": {"width":60,"depth":80,"height":28},
                    "features": ["landmark:%s"],
                    "params": { "module_id": "%s", "meshStructure": true, "designSeed": <vary 1-9999> } }
                Do NOT substitute MASS_MAIN for this named landmark.
                Still vary designSeed, facing, dimensions hints, and style_attributes within the style.

                """.formatted(moduleId, moduleId, moduleId);
        }
        return """

            ========================================
            LANDMARK MODULE ROUTING (RECOMMENDED — creative alternative allowed)
            ========================================
            User intent matches typology for module: %s
            PREFERRED path (high-fidelity bowl/ellipse):
              { "component_type": "MODULE",
                "relative_position": {"x":0,"y":0,"z":0},
                "dimensions": {"width":60,"depth":80,"height":28},
                "features": ["landmark:%s"],
                "params": { "module_id": "%s", "meshStructure": true, "designSeed": <vary 1-9999>,
                            "bowlSteepness": <0.25-0.45>, "facing": "SOUTH|NORTH|EAST|WEST" } }
            ALTERNATIVE (原创/独特): compositional MASS_MAIN (shape=circle|rounded_rect, masses[] tiers)
              + PAVING inner field + ROOF canopy + ENTRANCE — vary layout each time.
            Do NOT default to a plain rectangular MASS box for elliptical stadium requests.
            MassMain cannot render true elliptical bowl seating; use MODULE or plan_program + tiered masses.

            """.formatted(moduleId, moduleId, moduleId);
    }

    private static boolean containsAny(String lower, List<String> markers) {
        for (String m : markers) {
            if (m != null && !m.isBlank() && lower.contains(m.toLowerCase(Locale.ROOT))) {
                return true;
            }
        }
        return false;
    }

    private static boolean mentionsExplicitLandmark(String lower, LandmarkModuleRegistry.LandmarkModule module) {
        if (module.aliases() == null) return false;
        for (String alias : module.aliases()) {
            if (alias == null || alias.isBlank()) continue;
            String a = alias.toLowerCase(Locale.ROOT).trim();
            if (a.length() < 2) continue;
            // 跳过过宽的英文泛词，避免「stadium」误触发非鸟巢地标
            if (a.equals("stadium") || a.equals("arena") || a.equals("elliptical stadium")
                    || a.equals("oval stadium") || a.equals("体育馆") || a.equals("体育场")) {
                continue;
            }
            if (lower.contains(a)) {
                return true;
            }
        }
        return lower.contains(module.moduleId());
    }
}
